src/field_theory/primitive_element.lean

```diff
@@ -0,0 +1,327 @@
+Authors: Thomas Browning and Patrick Lutz
+-/
+
+import field_theory.adjoin
+import field_theory.separable
+
+/-!
+# Primitive Element Theorem
+
+In this file we prove the primitive element theorem.
+
+## Main results
+
+- `primitive_element`: a finite separable extension has a primitive element.
+
+-/
+
+noncomputable theory
+open_locale classical
+
+/- Proof of the primitive element theorem. -/
+
+open finite_dimensional
+
+namespace field
+
+section
+
+variables (F : Type*) [field F] {E : Type*} [field E] [algebra F E]
+
+/-! ### Primitive element theorem for finite fields -/
+
+-- Replaces earlier messy proof, courtesy of Aaron Anderson & Markus Himmel on zulip
+/-- A finite dimensional vector space over a finite field is finite. -/
+def finite_of_findim_over_finite [fintype F] [hE : finite_dimensional F E] : fintype E :=
+module.fintype_of_fintype (classical.some_spec (finite_dimensional.exists_is_basis_finset F E) : _)
+
+/-- Primitive element theorem assuming E is finite. -/
+lemma primitive_element_fin_aux [fintype E] : ∃ α : E, F⟮α⟯ = ⊤ :=
+begin
+  obtain ⟨α, hα⟩ := is_cyclic.exists_generator (units E),
+  use α,
+  ext,
+  rw iff_true_right algebra.mem_top,
+  by_cases hx : x = 0,
+  { rw hx,
+    exact F⟮α.val⟯.zero_mem, },
+  { obtain ⟨n, hn⟩ := set.mem_range.mp (hα (units.mk0 x hx)),
+    rw (show x = α^n, by norm_cast; simp *),
+    exact @is_subfield.pow_mem E _ α.val n F⟮α.val⟯ _ (field.mem_adjoin_simple_self F α.val), },
+end
+
+/-- Primitive element theorem for finite dimensional extension of a finite field. -/
+theorem primitive_element_fin [fintype F] [hfd : finite_dimensional F E] :
+  ∃ α : E, F⟮α⟯ = ⊤ :=
+begin
+  haveI : fintype E := finite_of_findim_over_finite F,
+  exact primitive_element_fin_aux F,
+end
+
+end
+
+section
+variables {F : Type*} [field F] {E : Type*} [field E] (ϕ : F →+* E)
+
+/-! ### Primitive element theorem for infinite fields -/
+
+lemma primitive_element_two_aux (α β : E) (f g : polynomial F) [F_inf : infinite F] :
+  ∃ c : F, ∀ (α' ∈ (f.map ϕ).roots) (β' ∈ (g.map ϕ).roots), β' ≠ β → ϕ c ≠ -(α' - α)/(β' - β) :=
+begin
+  let sf := (f.map ϕ).roots,
+  let sg := (g.map ϕ).roots,
+  let s := {c : E | ∃ (α' ∈ sf) (β' ∈ sg), β' ≠ β ∧ c = -(α' - α)/(β' - β)},
+  let r : E × E → E := λ ⟨α', β'⟩, -(α' - α)/(β' - β),
+  have hr : ∀ c ∈ s, ∃ α' β', ((α' ∈ sf) ∧ (β' ∈ sg)) ∧ r (α', β') = c,
+  { intros c hc,
+    rw set.mem_set_of_eq at hc,
+    tauto, },
+  have s_fin : s.finite,
+  { apply (set.finite.image r (set.finite_mem_finset (sf.to_finset.product sg.to_finset))).subset,
+    simpa only [set.subset_def, set.mem_image, prod.exists, multiset.mem_to_finset, finset.mem_product] using hr },
+  have s'_fin : (ϕ⁻¹' s).finite := s_fin.preimage ((ring_hom.injective ϕ).inj_on (⇑ϕ⁻¹' s)),
+  obtain ⟨c, hc⟩ := infinite.exists_not_mem_finset s'_fin.to_finset,
+  rw [set.finite.mem_to_finset, set.mem_preimage, set.mem_set_of_eq] at hc,
+  push_neg at hc,
+  exact ⟨c, hc⟩,
+end
```

Comment:
You can avoid a bit of pain in proving finiteness by sticking with `finset`s/`multiset`s. I also changed the conclusion slightly, which simplified `primitive_element_two_inv_key_aux` a tiny bit.
```suggestion
lemma primitive_element_two_aux (α β : E) (f g : polynomial F) [F_inf : infinite F] :
  ∃ c : F, ∀ (α' ∈ (f.map ϕ).roots) (β' ∈ (g.map ϕ).roots), -(α' - α)/(β' - β) ≠ ϕ c :=
begin
  let sf := (f.map ϕ).roots,
  let sg := (g.map ϕ).roots,
  let s := (sf.bind (λ α', sg.map (λ β', -(α' - α) / (β' - β)))).to_finset,
  let s' := s.preimage ϕ (λ x hx y hy h, ϕ.injective h),
  obtain ⟨c, hc⟩ := infinite.exists_not_mem_finset s',
  simp_rw [finset.mem_preimage, multiset.mem_to_finset, multiset.mem_bind, multiset.mem_map] at hc,
  push_neg at hc,
  exact ⟨c, hc⟩
end
```